From f100d2c332672acd7345e3240c6ead6177c5e8d9 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Tue, 28 Jun 2016 16:41:18 +0300
Subject: [PATCH 0301/1345] net: mvpp2x: add force link down and reset MAC to
 open/close port

commit  915991cbc59d865635b1433f8c768b4b4873fabe from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

-fix MAC reset issue
-show right status in ethtool

Change-Id: I9af9989316bc5da6bdc1980906c55784938a43fa
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30785
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
index 65d458a..c4f6ebc 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
@@ -882,7 +882,8 @@ int mv_gop110_port_init(struct gop_hw *gop, struct mv_mac_data *mac)
 
 	switch (mac->phy_mode) {
 	case PHY_INTERFACE_MODE_RGMII:
-		/* As workaround do not reset mac before MAC configuration */
+		mv_gop110_force_link_mode_set(gop, mac, false, true);
+		mv_gop110_gmac_reset(gop, mac_num, RESET);
 		/* configure PCS */
 		mv_gop110_gpcs_mode_cfg(gop, mac_num, false);
 
@@ -896,10 +897,13 @@ int mv_gop110_port_init(struct gop_hw *gop, struct mv_mac_data *mac)
 		mv_gop110_gpcs_reset(gop, mac_num, UNRESET);
 		/* mac unreset */
 		mv_gop110_gmac_reset(gop, mac_num, UNRESET);
+		mv_gop110_force_link_mode_set(gop, mac, false, false);
 	break;
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_QSGMII:
 		num_of_act_lanes = 1;
+		mv_gop110_force_link_mode_set(gop, mac, false, true);
+		mv_gop110_gmac_reset(gop, mac_num, RESET);
 		/* configure PCS */
 		mv_gop110_gpcs_mode_cfg(gop, mac_num, true);
 
@@ -912,6 +916,7 @@ int mv_gop110_port_init(struct gop_hw *gop, struct mv_mac_data *mac)
 		mv_gop110_gpcs_reset(gop, mac_num, UNRESET);
 		/* mac unreset */
 		mv_gop110_gmac_reset(gop, mac_num, UNRESET);
+		mv_gop110_force_link_mode_set(gop, mac, false, false);
 	break;
 	case PHY_INTERFACE_MODE_XAUI:
 		num_of_act_lanes = 4;
@@ -1052,6 +1057,8 @@ void mv_gop110_port_enable(struct gop_hw *gop, struct mv_mac_data *mac)
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_QSGMII:
 		mv_gop110_gmac_port_enable(gop, port_num);
+		mv_gop110_force_link_mode_set(gop, mac, false, false);
+		mv_gop110_gmac_reset(gop, port_num, UNRESET);
 	break;
 	case PHY_INTERFACE_MODE_XAUI:
 	case PHY_INTERFACE_MODE_RXAUI:
@@ -1075,6 +1082,8 @@ void mv_gop110_port_disable(struct gop_hw *gop, struct mv_mac_data *mac)
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_QSGMII:
 		mv_gop110_gmac_port_disable(gop, port_num);
+		mv_gop110_force_link_mode_set(gop, mac, false, true);
+		mv_gop110_gmac_reset(gop, port_num, RESET);
 	break;
 	case PHY_INTERFACE_MODE_XAUI:
 	case PHY_INTERFACE_MODE_RXAUI:
-- 
1.7.9.5

